src/librustc_span/hygiene.rs

```diff
@@ -848,14 +880,234 @@ impl DesugaringKind {
     }
 }
 
-impl Encodable for ExpnId {
-    fn encode<E: Encoder>(&self, _: &mut E) -> Result<(), E::Error> {
-        Ok(()) // FIXME(jseyfried) intercrate hygiene
+impl UseSpecializedEncodable for ExpnId {}
+impl UseSpecializedDecodable for ExpnId {}
+
+/// Additional information used to assist in decoding hygiene data
+pub struct HygieneContext {
+    // Maps serialized `SyntaxContext` ids to a `SyntaxContext` in the current
+    // global `HygieneData`. When we deserialize a `SyntaxContext`, we need to create
+    // a new id in the global `HygieneData`. This map tracks the ID we end up picking,
+    // so that multiple occurences of the same serialized id are decoded to the same
+    // `SyntaxContext`
+    remapped_ctxts: Lock<Vec<Option<SyntaxContext>>>,
+    // The same as `remapepd_ctxts`, but for `ExpnId`s
+    remapped_expns: Lock<Vec<Option<ExpnId>>>,
+}
+
+impl HygieneContext {
+    pub fn new() -> HygieneContext {
+        HygieneContext {
+            remapped_ctxts: Lock::new(Vec::new()),
+            remapped_expns: Lock::new(Vec::new()),
+        }
     }
 }
 
-impl Decodable for ExpnId {
-    fn decode<D: Decoder>(_: &mut D) -> Result<Self, D::Error> {
-        Ok(ExpnId::root()) // FIXME(jseyfried) intercrate hygiene
+pub fn decode_expn_id<
+    'a,
+    D: Decoder,
+    F: FnOnce(&mut D, u32) -> Result<ExpnData, D::Error>,
+    G: FnOnce(CrateNum) -> &'a HygieneContext,
+>(
+    d: &mut D,
+    mode: ExpnDataDecodeMode<'a, G>,
+    decode_data: F,
+) -> Result<ExpnId, D::Error> {
+    let index = u32::decode(d)?;
+    let context = match mode {
+        ExpnDataDecodeMode::IncrComp(context) => context,
+        ExpnDataDecodeMode::Metadata(get_context) => {
+            let krate = CrateNum::decode(d)?;
+            get_context(krate)
+        }
+    };
+
+    // Do this after decoding, so that we decode a `CrateNum`
+    // if necessary
+    if index == ExpnId::root().as_u32() {
+        debug!("decode_expn_id: deserialized root");
+        return Ok(ExpnId::root());
+    }
+
+    let outer_expns = &context.remapped_expns;
+
+    // Ensure that the lock() temporary is dropped early
```

Comment:
I believe these braces you have added here have no effect.

The current r-value lifetime rules will cause the lock() temporary to be dropped at the end of the evaluation of the expression; we do not extend it to the end of the containing block (which in this case would be the `fn` body of `decode_expn_id` itself).

(The place where you *do* need such braces is when you `let`-bind the value, as you do below with `let mut expns = outer_expns.lock();`.)

----

I would say that if your goal is to make the dynamic extent of the `lock()` temporary as clear as possible to future readers of this code, then I would keep the braces but I would also add a let-binding of the temporary, so that the braces are meaningful.

Reply:
Oh, here is a [playground demo](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d94962ae31ffe9dbe414251177b31ef9) illustrating what I mean when I say that these braces have no effect.

(The demo also points out a potentially surprising difference in the semantics of `if` vs `if let` in terms of what the r-value lifetime rules yield, but that is just a side-note; it does not contradict my main point that these braces have no effect.)

Reply:
> I would say that if your goal is to make the dynamic extent of the lock() temporary as clear as possible to future readers of this code, then I would keep the braces

That was my intention - I didn't want anyone reading the code to need to be aware of the exact rules for when temporaries are dropped.

> I would also add a let-binding of the temporary, so that the braces are meaningful.

I think the code looks cleaner as written - the important things is that the lock is dropped before calling `decode_data`, which the brackets make visually clear.

Reply:
I still think that if you want to make things explicit with braces,, then it should be braces that have an effect.

Putting braces in like this may lead people reading the code to think erroneously that these braces are *necessary* to encode the dynamic extent you desire here. I would prefer to avoid fueling such misunderstanding.

But given that the project does not have official guidance as to coding style in cases like this, I will not block landing the PR based on this disagreement.